cmd/vault-login-oauth/vault-login-oauth.go

```diff
@@ -0,0 +1,261 @@
+		//
+		// https://tools.ietf.org/html/rfc6819#section-5.3.5
+		// https://tools.ietf.org/html/rfc6749#section-10.12
+		// https://tools.ietf.org/html/rfc6749#section-4.1.2
+		fatal("DANGER: POSSIBLE CSRF ATTACK DETECTED.\nReceived state", state, "expected", login.stateNonce)
+		return
+	}
+
+	secret, err := login.vault.Write(login.pluginPath+"/login", map[string]interface{}{
+		"code":         code,
+		"redirect_uri": login.redirectURL,
+		"role":         login.role,
+	})
+
+	if err == nil {
+		writer.WriteHeader(http.StatusOK)
+		fmt.Fprint(writer, "authentication successful")
+		login.authErr = nil
+	} else {
+		writer.WriteHeader(http.StatusInternalServerError)
+		fmt.Fprint(writer, "authentication error: ", err)
+		login.authErr = err
+	}
+
+	login.authResponse = secret
+	login.authDone()
+}
+
+func (login *oidcLogin) runOAuthFlow() (*api.Secret, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), oauthTimeout)
+	login.authDone = cancel
+
+	// First we start the server.
+	go func() {
+		err := login.server.Serve(login.listener)
+		if err != nil && err != http.ErrServerClosed {
+			fatal("error running HTTP server:", err)
+		}
+	}()
+
+	// Then we open a browser.
+	openURL(login.authReqURL.String())
+
+	// handleOAuthCallback will cancel the context when the callback has been
+	// received, or the context will time out.
+	<-ctx.Done()
+	login.server.Shutdown(ctx)
+
+	if ctx.Err() != context.Canceled {
+		return nil, ctx.Err()
+	}
+	return login.authResponse, login.authErr
+}
+
+func main() {
+	// We must define our own flagSet because
+	// github.com/hashicorp/vault/command/token depends on "testing", which
+	// will pollute the global flags :(
+	flagSet := flag.NewFlagSet("vault-login-oauth", flag.ExitOnError)
+	roleFlag := flagSet.String("role", "default", "Authenticate as this role")
+	pathFlag := flagSet.String("path", "/auth/oauth", "Mount point of the OAuth plugin")
+
+	flagSet.Parse(os.Args[1:])
+
+	login := newLogin(*pathFlag, *roleFlag)
+	login.makeVaultClient()
+	login.startListening()
+	login.getAuthReqURL()
+
+	response, err := login.runOAuthFlow()
+	if err == context.DeadlineExceeded {
+		fatal("Timed out waiting for response")
+	} else if err != nil {
+		fatal(err)
+	}
+
+	if response == nil || response.Auth == nil {
+		fatal("Response contained no authentication information")
+	}
+
+	clientToken := response.Auth.ClientToken
+	if clientToken == "" {
+		fatal("No token in response!")
+	}
+	tokenHelper := token.InternalTokenHelper{}
+	err = tokenHelper.Store(clientToken)
+	if err != nil {
+		fatal("Failed to save token:", err)
+	}
+	os.Exit(0)
+}
+
+// fatal prints something to stderr and exits immediately.
+func fatal(v ...interface{}) {
+	fmt.Fprintln(os.Stderr, v...)
+	os.Exit(1)
+}
+
+func openURL(url string) {
+	var err error
+
+	switch runtime.GOOS {
+	case "linux":
+		err = exec.Command("xdg-open", url).Start()
+	case "windows":
+		err = exec.Command("rundll32", "url.dll,FileProtocolHandler", url).Start()
```

Comment:
I think this is essentially just a ShellExecute, which will run any arbitrary executable that is local or accessible via SMB/WebDav.  Ex:
rundll32 url.dll,FileProtocolHandler file://c:\windows\notepad.exe
...launches a local Notepad.exe instance when executed from the command line on Windows.

While there is some validation logic in getAuthReqURL, it probably doesn't effectively prevent code execution.

The url comes from vault so presumably it can be trusted, but still it's prudent to avoid ShellExecuting the URL when all we really need is to load web page.
If we don't immediately see a safe way to _just_ launch the local browser aimed at a URL, at minimum it makes sense to validate the URL scheme is https:// in openURL.

Not sure if this also applies to the Linux & Darwin exec mechanisms.